llvmPackages_14: Fix build on aarch64-linux

avdv opened the pull request:
###### Description of changes

This PR fixes building the LLVM packages on aarch64-linux.

The libcxx library needs a C++20 capable compiler toolchain since version 14.

Also, work around the linking problems described in https://github.com/NixOS/nixpkgs/issues/201254

ZHF: #199919

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

risicle commented:
@ofborg eval

avdv commented on pkgs/development/compilers/llvm/14/llvm/default.nix:
```diff
@@ -129,6 +129,9 @@ in stdenv.mkDerivation (rec {
 
+  # workaround https://github.com/NixOS/nixpkgs/issues/201254
+  NIX_LDFLAGS = optionalString (stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU) "-lgcc";
+
```
I suppose this change might be the reason that this would cause so many rebuilds, even on Darwin which should really not be effected at all.

Should I change that to add the `NIX_LDFLAGS` attribute optionally?

vcunat replied:
You can avoid rebuilds by using
```nix
{
  NIX_LDFLAGS = if stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU then "-lgcc" else null;
}
```

avdv replied:
Thank you, @vcunat! Changed.

risicle commented:
Hmm I'm still getting

```
/nix/store/1g5g7vzlcprmg1d6864fs5r4nf1h5b5j-binutils-2.39/bin/ld: CMakeFiles/clang_rt.scudo_standalone-dynamic-aarch64.dir/linux.cpp.o: in function `bool scudo::atomic_compare_exchange_strong<scudo::atomic_u32>(scudo::atomic_u32 volatile*, scudo::atomic_u32::Type*, scudo::atomic_u32::Type, scudo::memory_order)':
/build/compiler-rt-src-14.0.6/compiler-rt/lib/scudo/standalone/atomic_helpers.h:119: undefined reference to `__aarch64_cas4_acq'
```

for `compiler-rt-libc-14.0.6`

avdv commented:
@risicle you're right, I have added the same workaround for compiler-rt and clang.

avdv commented:
When building `llvmPackages_14.libcxxabi`, I ran into this error:
```
n file included from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/strong_order.h:12,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/weak_order.h:14,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/partial_order.h:14,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/compare_partial_order_fallback.h:13,                   
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/compare:144,                                         
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/utility:236,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__functional_base:26,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/memory:808,         
                 from /build/libcxxabi-src-14.0.6/libcxxabi/../libcxx/src/include/atomic_support.h:13,
                 from /build/libcxxabi-src-14.0.6/libcxxabi/src/cxa_default_handlers.cpp:19:
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h: In function 'constexpr _ToType std::__1::bit_cast(const _FromType&)':
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h:31:38: error: expected primary-expression before ',' token
   31 |     return __builtin_bit_cast(_ToType, __from);
      |                                      ^                                                                                                                                   
In file included from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/strong_order.h:12,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/weak_order.h:14,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/partial_order.h:14,                                             
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/compare_partial_order_fallback.h:13,                
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/compare:144,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/utility:236,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__functional_base:26,        
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/memory:808,
                 from /build/libcxxabi-src-14.0.6/libcxxabi/../libcxx/src/include/atomic_support.h:13,
                 from /build/libcxxabi-src-14.0.6/libcxxabi/src/cxa_default_handlers.cpp:19:         
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h: In function 'constexpr _ToType std::__1::bit_cast(const _FromType&)':
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h:31:38: error: expected primary-expression before ',' token
   31 |     return __builtin_bit_cast(_ToType, __from);                                                                                                                          
      |                                      ^                                                                                                                                   
make[2]: *** [src/CMakeFiles/cxxabi_static.dir/build.make:90: src/CMakeFiles/cxxabi_static.dir/cxa_default_handlers.cpp.o] Error 1               
make[1]: *** [CMakeFiles/Makefile2:165: src/CMakeFiles/cxxabi_static.dir/all] Error 2                                                                                            
make[1]: *** Waiting for unfinished jobs....                                                                                                                                     
make[2]: *** [src/CMakeFiles/cxxabi_shared.dir/build.make:90: src/CMakeFiles/cxxabi_shared.dir/cxa_default_handlers.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....                                                                    
```
According to https://github.com/llvm/llvm-project/issues/56828 this happens with GCC 10, but works OK with GCC 11.

risicle commented:
Linking error again for `lldb`

ehmry commented on pkgs/development/compilers/llvm/14/clang/default.nix:
```diff
@@ -26,6 +26,9 @@ let
     buildInputs = [ libxml2 libllvm ];
 
+    # workaround https://github.com/NixOS/nixpkgs/issues/201254
+    NIX_LDFLAGS = if stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU then "-lgcc" else null;
```
We should consider if adding an argument above like `{ withLibgcc ? stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU, ... }:` is better than putting these conditions inline in the body of derivations. This could make overriding LLVM easier and I think its self-documentation to have code in the form of `effect ? cause` rather than `if ${cause} then ${action} else null`.

avdv replied:
Since this is just an ugly workaround, I would prefer to keep this as ugly as possible. :grimacing:

avdv commented:
> Linking error again for `lldb`

Fixed.

I tried to use nixpkgs-review, but this never finishes... :shrug:

primeos commented on pkgs/development/compilers/llvm/14/clang/default.nix:
```diff
@@ -1,6 +1,14 @@
-{ lib, stdenv, llvm_meta
-, monorepoSrc, runCommand
-, substituteAll, cmake, libxml2, libllvm, version, python3
+{ lib
+, stdenv
+, llvm_meta
+, monorepoSrc
+, runCommand
+, substituteAll
+, cmake
+, libxml2
+, libllvm
+, version
+, python3
```
I'm not a fan of reformatting code in commits like `llvmPackages_14: Add -lgcc to NIX_LDFLAGS for lldb`. IMO that belongs in a separate commit (especially since https://github.com/NixOS/nixpkgs/pull/201485/commits/b93b4f60855d15fc7784b5fc5c57cb0c3cd899b2 doesn't seem to make any functional changes in this file at all so maybe this change was added by accident?).

avdv replied:
Yes, this was added by accident. Auto-formatted file + `git commit -am ...`.

avdv replied:
fixed.

primeos reviewed (changes requested):
Just as a TODO: The commit history needs to be cleaned up (one commit per logical change - currently there are basically a few fixup commits - I'd suggest to either put everything into one commit or one to fix the build and one for the `-lgcc` changes (the latter would probably be better)).

Ideally this should also target `llvmPackages_git` as well (normally that would be required but `llvmPackages_git` is in bad shape since many PRs forget to target it and `llvmPackages_15` already has a PR that is stuck in limbo...).

avdv committed:
llvmPackages_14: Use GCC 11 on aarch64-linux

The libcxx library needs a C++20 capable compiler toolchain since version 14, but aarch64-linux uses GCC 9 by default.

However, building libcxxabi fails when using GCC 10, see [llvm/llvm-project#56828]. So let's use GCC 11 instead.

[llvm/llvm-project#56828]: https://github.com/llvm/llvm-project/issues/56828

avdv committed:
llvmPackages_14: Add -lgcc to NIX_LDFLAGS to lld, lldb, llvm, compiler-rt and clang

This is required to workaround #201254.

avdv commented:
@primeos  I have rebased onto master and refined the commit history.

primeos reviewed (approved):
LGTM now, thanks! :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/694

vcunat commented:
This can't go directly to master.  OfBorg doesn't check this for aarch64-linux, but it is a large rebuild:
```
Estimating rebuild amount by counting changed Hydra jobs (parallel=unset).
  14523 aarch64-linux
```

mweinelt commented:
Bisected a broken `llvm_14` on aarch64-linux to 74965741696bf79fd67501a55b1b60683c2b5d62.

In https://hydra.nixos.org/build/201827173 it fails >80 tests.

ehmry commented:
Go ahead and revert my merge if it doesn't make things worse.

Artturin commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14749,6 +14749,8 @@ with pkgs;
     buildLlvmTools = buildPackages.llvmPackages_14.tools;
     targetLlvmLibraries = targetPackages.llvmPackages_14.libraries or llvmPackages_14.libraries;
     targetLlvm = targetPackages.llvmPackages_14.llvm or llvmPackages_14.llvm;
+  } // lib.optionalAttrs (stdenv.hostPlatform.isAarch64 && stdenv.hostPlatform.isLinux && buildPackages.stdenv.cc.isGNU) {
+    stdenv = gcc11Stdenv;
```
https://github.com/NixOS/nixpkgs/issues/108305

avdv commented:
> Bisected a broken `llvm_14` on aarch64-linux to [7496574](https://github.com/NixOS/nixpkgs/commit/74965741696bf79fd67501a55b1b60683c2b5d62).
> 
> In [hydra.nixos.org/build/201827173](https://hydra.nixos.org/build/201827173) it fails >80 tests.

It fails all the gold tests. Which is caused by:
```
/nix/store/cwivsnh1p591x7yydynrz0xc1zcq54p3-binutils-2.39/bin/ld.gold: error: /build/llvm-src-14.0.6/llvm/build/lib/LLVMgold.so: could not load plugin library: /nix/store/kyrb20x8ldfhih953ffd8yba42qfb5xy-gcc-9.5.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /build/llvm-src-14.0.6/llvm/build/lib/libLLVM-14.so)
```
Would it be feasible to disable the gold linker plugin on this platform?
